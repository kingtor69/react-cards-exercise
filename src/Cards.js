import React, {useState, useEffect} from 'react';
import axios from 'axios';
import Card from './Card';

const Cards = ({deckBaseUrl, deckNewEnd, deckDrawEnd}) => {
  const [numDecks, setNumDecks] = useState(1);
  const [numCards, setNumCards] = useState(1);
  const [deckId, setDeckId] = useState(null);
  const [cardsRemaining, setCardsRemaining] = useState(0);
  const [dealtCards, setDealtCards] = useState([]);

  const shuffleNewDeck = async () => {
    const res = await axios.get(`${deckBaseUrl}${deckNewEnd}${numDecks}`);
    setDeckId(res.data.deck_id);
    setCardsRemaining(res.data.remaining);
  };

  const dealCard = async () => {
    const res = await axios.get(`${deckBaseUrl}${deckId}${deckDrawEnd}${numCards}`);
    setCardsRemaining(res.data.remaining);
    setDealtCards(dealtCards => [...dealtCards, ...res.data.cards]);
  };

  useEffect(() => {
    shuffleNewDeck();
  }, []);

  const fewerCards = () => {
    if (numCards > 1) {
      setNumCards(numCards -1)
    };
  };

  const moreCards = () => {
    if (numCards < cardsRemaining) {
      setNumCards(numCards +1)
    };
  };

  return (
    <div className="Cards">
      <h2>Let's throw some cards around, eh?</h2>
      {(cardsRemaining > 0) 
        ? <>
            <button 
              onClick={dealCard} 
              className="big-btn"
              data-testid="deal"
            >
              deal {numCards} card{(numCards > 1) ? 's' : ''}
            </button> 
            <br />
            <button 
              onClick={fewerCards}
              data-testid="fewer-cards"
            >
              fewer cards
            </button>
            <button 
              onClick={moreCards}
              data-testid="more-cards"
            >
              more cards
            </button>
          </>
        : <>
            <button 
              onClick={shuffleNewDeck} className="big-btn"
              data-testid="shuffle"
            >
              shuffle {numDecks} deck{(numDecks > 1) ? 's' : ''}
            </button> 
            <br />
            <button 
              onClick={() => {setNumDecks(numDecks -1)}}
              data-testid="fewer-decks"
            >
              fewer decks
            </button>
            <button 
              onClick={() => {setNumDecks(numDecks +1)}}
              data-testid="more-decks"
            >
              more decks
            </button>
          </>
      }
      <div className="table">
        {dealtCards.map((card, i) => (
            <Card 
              img={card.image}
              code={card.code}
              offset={i*20}
              key={i} 
            />
        ))}
      </div>
    </div>
  );
};

export default Cards;