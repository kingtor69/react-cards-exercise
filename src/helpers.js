const deckBaseUrl = "http://deckofcardsapi.com/api/deck/"
const deckNewEnd = "new/shuffle/?deck_count="
const deckDrawEnd = "/draw/?count="

export {deckBaseUrl,
        deckNewEnd,
        deckDrawEnd};