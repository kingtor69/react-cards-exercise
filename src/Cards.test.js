import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Cards from './Cards';
import {deckBaseUrl, deckNewEnd, deckDrawEnd} from './helpers';

it('renders without crashing', () => {
  render(<Cards />);
});

it('renders snapshot', () => {
  const { asFragment } = render(<Cards />);
  expect(asFragment()).toMatchSnapshot();
});

it('should change cards dealt with "fewer" and "more" buttons', () => {
  const { queryByText, queryByTestId } = render(<Cards 
    deckBaseUrl={deckBaseUrl}
    deckNewEnd={deckNewEnd}
    deckDrawEnd={deckDrawEnd}
  />);
  const more = queryByText('more cards');
  const fewer = queryByText('fewer cards');
  const deal = queryByTestId('deal');
  expect(more).toBeInTheDocument();
  expect(fewer).toBeInTheDocument();
  expect(deal).toBeInTheDocument();
  // expect(queryByText("deal 1 card")).toBeInTheDocument();
  // expect(queryByText("deal 2 cards")).toBe(null);
  // fireEvent.click(more);
  // expect(queryByText("deal 2 cards")).toBeInTheDocument();
  // fireEvent.click(fewer);
  // expect(queryByText("deal 2 cards")).toBe(null);
  // expect(queryByText("deal 1 card")).toBeInTheDocument();
})