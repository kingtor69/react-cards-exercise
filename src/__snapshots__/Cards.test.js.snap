// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="Cards"
  >
    <h2>
      Let's throw some cards around, eh?
    </h2>
    <button
      class="big-btn"
      data-testid="shuffle"
    >
      shuffle 1 deck
    </button>
    <br />
    <button
      data-testid="fewer-decks"
    >
      fewer decks
    </button>
    <button
      data-testid="more-decks"
    >
      more decks
    </button>
    <div
      class="table"
    />
  </div>
</DocumentFragment>
`;
